### PR TITLE
Add tests to verify flow validation negative cases

### PR DESCRIPTION
**Pull request by npliashechnikov**

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        assert Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow.id).status) }
+        def forward = database.getFlow(flow.id)?.left
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id).collect {it -> it.dpId}
+
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
+
+        when: "${rule} switch rule is deleted"
+        def cookie = newRules[newRules.keySet()[item]][0].cookie.toString()
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Omg it literally took me about 40 minutes to understand that this `cookie` is the same as `forward.cookie`. I don't understand why are you trying to obfuscate the simple test.
```
given: "Flow with transit switch"
//everything fine here until assert that flow is up at line 74

and: "Defined forward cookie to be deleted"
def fwCookie = database.getFlow(flow.id).left.cookie

and: "The cookie is present on all switches of the flow"
//by the way I don't think that this verification should be present in this test. Probably should be present 
//in flow crud, but not here. Still adding it to match your original steps
def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)
assert involvedSwitches.every{ sw ->
   northbound.getSwitchRules(sw.dpId).flowEntries.find { it.cookie == fwCookie }
}

when: "Delete #rules cookie"
northbound.deleteSwitchRules(involvedSwitches[item], fwCookie)

then: "Flow validation reports missing cookie"
def validation = northbound.validateFlow(flow.id)
//verify that only forward flow is broken
//verify that only fwCookie is reported as missing 
??profit
```
This is it. No need for additional mind blowing `Map<SwitchId, List<FlowEntry>> findFlowRules` methods and obfuscated test code

File: services/src/functional-tests/src/test/resources/topology.yaml [hunk 2]
```diff
@@ -1,57 +1,57 @@
 switches:
     - name: ofsw1
-      dp_id: 00:00:00:00:00:00:00:01
+      dp_id: 00:00:00:22:3d:6b:00:04
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/resources/topology.yaml]: This should not be pushed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 1 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: If you want we can discuss how this implementation is different from yours and weigh the pros and cons

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        Map<SwitchId, List<FlowEntry>> result = [:]
+        captureSwitchRules(switches).findAll { k, v -> result[k] = v.flowEntries.findAll { it.cookie == cookie } }
+        return result
+    }
+
+    def noDirectLinks(Switch src, Switch dst, List<IslInfoData> links) {
+        def connectingLinks = links.findAll {link ->
+            (link.path[0].switchId == src.dpId) && (link.path[-1].switchId == dst.dpId)
+        }
+        return connectingLinks == []
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    Map<SwitchId, SwitchFlowEntries> captureSwitchRules(List<SwitchId> switchIds) {
+        def rules = [:]
+        switchIds.each { switchId -> rules[switchId] = northbound.getSwitchRules(switchId) }
+        return rules
+    }
+
+    Map<SwitchId, List<FlowEntry>> findFlowRules(long cookie, List<SwitchId> switches) {
+        Map<SwitchId, List<FlowEntry>> result = [:]
+        captureSwitchRules(switches).findAll { k, v -> result[k] = v.flowEntries.findAll { it.cookie == cookie } }
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Replaced findAll to each + changed param names

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        return result
+    }
+
+    def noDirectLinks(Switch src, Switch dst, List<IslInfoData> links) {
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Changed to checkDirectLinks()

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        def switches = topologyDefinition.getActiveSwitches()
+        return [switches.first(), switches.first()]
+    }
+
+    @Unroll
+    def "Flow validation should fail if #rule switch rule is deleted"() {
+        given: "Flow with a transit switch exists"
+
+        def (Switch srcSwitch, Switch dstSwitch) = getNonNeighbouringSwitches()
+        assert srcSwitch && dstSwitch
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Removed check

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
+
+        when: "${rule} switch rule is deleted"
+        def cookie = newRules[newRules.keySet()[item]][0].cookie.toString()
+        northbound.deleteSwitchRules(newRules.keySet()[item], cookie)
+        then: "Flow validation should fail"
+
+        !northbound.validateFlow(flow.id).every {it -> it.asExpected}
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Reimplemented flow validation checks.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id).collect {it -> it.dpId}
+
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
+
+        when: "${rule} switch rule is deleted"
+        def cookie = newRules[newRules.keySet()[item]][0].cookie.toString()
+        northbound.deleteSwitchRules(newRules.keySet()[item], cookie)
+        then: "Flow validation should fail"
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Would like to keep this as is - would have to change steps naming scheme throughout all test suite

[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: np, then you'll have to change steps naming scheme throughout whole `services/src/functional-tests` project except for this spec

File: services/src/functional-tests/src/test/resources/topology.yaml (hunk 2 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/resources/topology.yaml]: Reverted, thanks

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Nice idea btw - from my experience with multiple project and test frameworks, passive voice in the requirements and the test steps descriptions is much more common.

[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: ok, I just noted how it is done here. Let's leave as is, I don't think that's too important right now. We can do sth about it later. First of all it should be readable, and it *is* in both options

File: services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java [hunk 9]
```diff
@@ -59,7 +61,11 @@ public DatabaseNeoImpl(@org.springframework.beans.factory.annotation.Value("${ne
     @Override
     public void destroy() {
         if (neo != null) {
-            neo.close();
+            try {
+                neo.close();
+            } catch (Exception e) {
```
Comment (rtretyak) [inline services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java]: Why do we need this code? I don't understand why we should catch exceptions in test code. Did you face any issues here?

File: services/src/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundService.java [hunk 10]
```diff
@@ -75,6 +75,8 @@
 
     List<Long> deleteSwitchRules(SwitchId switchId, DeleteRulesAction deleteAction);
 
+    List<Long> deleteSwitchRules(SwitchId switchId, String cookie);
```
Comment (rtretyak) [inline services/src/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundService.java]: Should be `long cookie` not `String`

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        assert involvedSwitches.every{ sw ->
+            northbound.getSwitchRules(sw).flowEntries.find { it.cookie == forward.cookie }
+        }
+        when: "${rule} switch rule is deleted"
+        def dpId = involvedSwitches[item]
+        def cookie = forward.cookie.toString()
+        northbound.deleteSwitchRules(dpId, cookie)
+
+        def validationResult = northbound.validateFlow(flow.id)
+        then: "Forward and reverse flows should be validated"
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        validationResult.size() == 2
+        and: "Reverse flow validation must be successful"
+        validationResult.any { isFlowValid(it) }
+        and: "Forward flow validation should fail"
+        def fwdFlow = validationResult.find { !isFlowValid(it) }
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Done

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+        if (flow){
+            northbound.deleteFlow(flow.id)
+        }
+        where:
+        rule << ["ingress", "transit", "egress"]
+        item << [0, 1, -1]
+    }
+
+    @Unroll
+    def "Both flow and switch validation should fail if flow rule is missing with #flowconfig flow configuration"() {
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: First test covers only non-neighboring switches, second one tries all 3 possible flow configurations. It is possible to factor 1st test case in such a way, but I'm afraid it'd get even more messy.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+        def flow2 = flowHelper.singleSwitchFlow(src)
+        flow2 = northbound.addFlow(flow2)
+        assert flow2?.id
+        Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow2.id).status) }
+
+        if (flowType == "forward") {
+            damagedFlow = database.getFlow(flow1.id).left
+        } else {
+            damagedFlow = database.getFlow(flow1.id).right
+        }
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Will use first approach, thanks

File: services/src/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundService.java (hunk 10 above)
Comment (npliashechnikov) [reply services/src/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundService.java]: Fixed

File: services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java (hunk 9 above)
Comment (npliashechnikov) [reply services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java]: I/O exceptions in destructors are not good. Decided to avoid it anyway.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        Flow forward = database.getFlow(flow.id).left
+
+        when: "A flow rule gets deleted"
+        northbound.deleteSwitchRules(involvedSwitches[item], forward.cookie.toString())
+
+        then: "Flow and switch validation should fail"
+        northbound.validateFlow(flow.id).findAll {isFlowValid(it)} .size() == 1
+        northbound.validateFlow(flow.id).findAll {!isFlowValid(it)} .size() == 1
+
+        northbound.validateSwitchRules(involvedSwitches[item]).missingRules.size() > 0
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Added the verifications, thanks

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+        when: "${rule} switch rule is deleted"
+        def dpId = involvedSwitches[item]
+        def cookie = forward.cookie.toString()
+        northbound.deleteSwitchRules(dpId, cookie)
+
+        def validationResult = northbound.validateFlow(flow.id)
+        then: "Forward and reverse flows should be validated"
+        validationResult.size() == 2
+        and: "Reverse flow validation must be successful"
+        validationResult.any { isFlowValid(it) }
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: I'm genuinely unsure if we can safely assume that.

File: services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java (hunk 9 above)
Comment (npliashechnikov) [reply services/src/test-library/src/main/java/org/openkilda/testing/service/database/DatabaseNeoImpl.java]: Removed the code.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+    @Unroll
+    def "Flow validation should fail if #rule switch rule is deleted"() {
+        given: "Flow with a transit switch exists"
+
+        def (Switch srcSwitch, Switch dstSwitch) = getNonNeighbouringSwitches()
+        assert srcSwitch && dstSwitch
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow = northbound.addFlow(flow)
+        assert flow?.id
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        assert Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow.id).status) }
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+        def forward = database.getFlow(flow.id)?.left
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id).collect {it -> it.dpId}
+
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
+
+        when: "${rule} switch rule is deleted"
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Looks like Spock parses that.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow = northbound.addFlow(flow)
+        assert flow?.id
+        assert Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow.id).status) }
+        def forward = database.getFlow(flow.id)?.left
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id).collect {it -> it.dpId}
+
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Replaced by looking up the rule cookie on each switch.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Replaced by using cookie from Neo

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+        def validationResult = northbound.validateFlow(flow.id)
+        then: "Forward and reverse flows should be validated"
+        validationResult.size() == 2
+        and: "Reverse flow validation must be successful"
+        validationResult.any { isFlowValid(it) }
+        and: "Forward flow validation should fail"
+        def fwdFlow = validationResult.find { !isFlowValid(it) }
+        and: "Only one rule should be missing"
+        def rules = findRulesDiscrepancies(fwdFlow)
+        rules.keySet().size() == 1
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+
+    @Autowired
+    PathHelper pathHelper
+
+    @Autowired
+    Database database
+
+
+    boolean isFlowValid(FlowValidationDto flow, boolean ignoreMeters=true) {
+        if (ignoreMeters) {
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Fixed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+        }
+        where:
+        rule << ["ingress", "transit", "egress"]
+        item << [0, 1, -1]
+    }
+
+    @Unroll
+    def "Both flow and switch validation should fail if flow rule is missing with #flowconfig flow configuration"() {
+        given: "A ${flowconfig} flow exists"
+        def (src, dest) = method()
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Take a look at "where" block.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Cases of deleting ingress/egress rules on flow with transit switches will be omitted if I delete first test. As I said, it could be merged into a single test cases, but what for?

[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: > Cases of deleting ingress/egress rules on flow with transit switches will be omitted if I delete first test
true, that makes sense. So does the 'transit' case is duplicating or what is the difference?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 23 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Renamed "method" to "switchSelectMethod", is it acceptable now?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: So what if add to the where block next values
```
        where:
        flowconfig << ["ingress on flow with transit switch", "egress on flow with transit switch"]
        method << [{ getNonNeighbouringSwitches() }, { getNonNeighbouringSwitches() }]
        item << [0, -1]
```
Will it fully cover the 1st test or what are the other differences?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 23 above)
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Yes, it is better. I don't remember if I already asked this, but why are we wrapping the `{ getNonNeighbouringSwitches() }` into closures in 'where' block? I understand that this way we delay the calculation of switchPairs until the test run, but I don't understand why. From my point it should work fine without `{ }`.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+
+        then: "Forward and reverse flows should be validated"
+        validationResult.size() == 2
+
+        and: "Reverse flow validation must be successful"
+        def reverseResult = validationResult.find { isFlowValid(it) }
+        !reverseResult.toString().contains(cookie.toString())
+
+        and: "Forward flow validation should fail"
+        def forwardResult = validationResult.find { !isFlowValid(it) }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: This block itself is not performing verification. Note that `find` method will return `null` if no results found (will not throw or sth), so you should add
```
and: "Forward flow validation should fail"
def forwardResult = validationResult.find { !isFlowValid(it) }
forwardResult
```
to verify that forwardResult is actually found. Otherwise you may just assign 'null' to forwardResult and continue test execution.  
Same in the above block, but at least it will throw a NPE at line 113 (which is still not the best way to fail the test, though still viable)

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 13 above)
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Resolved by factoring the test cases

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+        switchValidationResult.missingRules[0] == damagedFlow.cookie
+
+        and: "There should be no excess rules on the affected switch"
+        switchValidationResult.excessRules.size() == 0
+
+        and: "Validation of non-affected switches (if any) should succeed"
+        if (damagedFlowSwitches.size() > 1) {
+            def nonAffectedSwitches = damagedFlowSwitches.findAll { it != damagedFlowSwitches[item] }
+            nonAffectedSwitches.every { sw -> northbound.validateSwitchRules(sw).missingRules.size() == 0 }
+            nonAffectedSwitches.every { sw -> northbound.validateSwitchRules(sw).excessRules.size() == 0 }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Since you are inside the `if` statement, Spock's implicit assertion will not work here. So you should explicitly put `assert` keywords here, otherwise this is not being verified

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        and: "Delete the flows"
+        [flowToBreak?.id, intactFlow?.id].each { deleteFlow(it) }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: your safe navigation here makes very little sense. If, for example, intact flow will be null, then you'll loop over
`[flowtoBreakId, null]. each { deleteFlow(it) }` and will still pass null to the deleteFlow method which will anyway result in NPE.  
Also, your safe navigation makes no sense, because previously in the test you already accessed `flowToBreak.id` without safe navigation. Since this is not a `cleanup` block, this code will only get executed if previous code worked, so..

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy [hunk 27]
```diff
@@ -107,7 +107,17 @@ class PathHelper {
...
+        return (List<Switch>)getInvolvedIsls(path).collect { [it.srcSwitch, it.dstSwitch] }.flatten().unique()
+    }
+
+    /**
+     * Get list of Switches involved in given flow.
+     */
+    List<Switch> getInvolvedSwitches(String flowId) {
+        def flowPath = northbound.getFlowPath(flowId)
+        assert flowPath.forwardPath
+        def switchIds = flowPath.forwardPath.unique()*.switchId
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy]: what does the `forwardPath.unique()` do?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 26 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Makes sense, thanks

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy (hunk 27 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy]: It is not possible

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 28]
```diff
@@ -0,0 +1,205 @@
...
+    @Autowired
+    Database database
+
+    /**
+     * Checks if there is no discrepancies in the flow validation results
+     * //TODO: Don't skip MeterId discrepancies when virtual env starts supporting the meters.
+     * @param flow Flow validation results
+     * @return boolean
+     */
+    boolean isFlowValid(FlowValidationDto flow) {
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Please, move all helper methods under tests. It is inconvenient to scroll down a lot to find the test.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy [hunk 29]
```diff
@@ -0,0 +1,205 @@
...
+     * Finds a single switch
+     * @return List<Switch> (convenient for randomFlow() method)
+     */
+    def getSingleSwitch() {
+        def switches = topologyDefinition.getActiveSwitches()
+        return [switches.first(), switches.first()]
+    }
+
+    //TODO: Use FlowHelper's deleteFlow() as it gets merged
+    def deleteFlow(String flowId) {
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: FlowHelper's deleteFlow() is merged.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy [hunk 30]
```diff
@@ -107,7 +107,17 @@ class PathHelper {
...
      */
     List<Switch> getInvolvedSwitches(List<PathNode> path) {
-        return getInvolvedIsls(path).collect { [it.srcSwitch, it.dstSwitch] }.flatten().unique()
+        return (List<Switch>)getInvolvedIsls(path).collect { [it.srcSwitch, it.dstSwitch] }.flatten().unique()
+    }
+
+    /**
+     * Get list of Switches involved in given flow.
+     */
+    List<Switch> getInvolvedSwitches(String flowId) {
```
[the remark this answers is not in the document]
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy]: It was confirmed by @rtretyak that we're going to eventually get rid of convert() method and the related stuff.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy (hunk 29 above)
Comment (npliashechnikov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy]: Thanks, will use now

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy [hunk 31]
```diff
@@ -107,7 +107,14 @@ class PathHelper {
      * Get list of Switches involved in given path.
      */
     List<Switch> getInvolvedSwitches(List<PathNode> path) {
-        return getInvolvedIsls(path).collect { [it.srcSwitch, it.dstSwitch] }.flatten().unique()
+        return (List<Switch>)getInvolvedIsls(path).collect { [it.srcSwitch, it.dstSwitch] }.flatten().unique()
+    }
+
+    /**
+     * Get list of Switches involved in an existing/UP flow
```
Comment (ylobankov) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/PathHelper.groovy]: You don't check that the flow is UP.
